Replace Python core with pure-Vimscript core ("vimcore")

cxw42 opened the pull request:
Here it is - let me know what you think!  Passes all tests except Windows UTF-8, which @ppalaga says might not be possible (https://github.com/editorconfig/editorconfig-core-c/pull/31#issue-154810185).

Closes https://github.com/editorconfig/editorconfig/issues/383 .
Closes #48 (replaces it).

Also closes the open Python issues:
Closes #57.
Closes #101.
Closes #116.

May help with #108.
May affect #111.

Part of the discussion in https://github.com/vim/vim/issues/2286 .

cxw42 commented:
The CI failure appears to be related to rake/bundler/vimrunner versions.  Would someone with some Ruby expertise be able to give me some suggestions?

xuhdev commented:
I guess you have generated the lock file in a new version of bundler. The easiest way to resolve this is probably use bundler less than 2.0 to generate Gemfile.lock on your system.

cxw42 commented:
@xuhdev Thanks - that worked!  I was able to run the tests on Cygwin from `editorconfig-vim/` with

    export BUNDLE_GEMFILE=tests/Gemfile
    bundle _1.15.4_ exec rspec tests/spec/editorconfig_spec.rb

Please let me know if you would like any changes, or if you have any concerns.

cxw42 commented:
@xuhdev Any thoughts?  I am wondering if we can merge this one.  After that, I would like to work with @rakus to see if we can merge some of his plugin changes into this official plugin.  I think removing the Python should come before any Vimscript enhancements.  What say you?

xuhdev commented:
I remember you said something about number range (`{-3...100}`) and this core depends on that. I wanted to start a voting for that I couldn't find that thread. Could you remind me where it is?

cxw42 commented:
@xuhdev 

> CI

The CI (Travis) is unchanged in this PR.  The core is copied from https://github.com/cxw42/editorconfig-core-vimscript , which has its own CI (Appveyor) set up.

> number range

The number range is editorconfig/editorconfig#371 .  Thank you for following up on it!

I changed this PR so it does not depend on that issue.  It has the current behaviour (e.g., `{-10..10}` rejects `0`).  Therefore, I think this can go ahead even before the vote on number ranges.

xuhdev commented:
I understand you haven't changed anything about CI. What I meant is that, if you add a core library to the repo, we need CI to test that core library for guaranteeing its continuously correct functioning...

cxw42 commented:
@xuhdev Sorry for misunderstanding!  I have not but am happy to.  However, the vimscript core requires a number of support files in order to be testable by ctest (e.g., bash scripts and bat files).  May I add a separate repo submodule for those so plugin users don't get files they don't need?

xuhdev commented:
I personally feel it's OK to add editorconfig/editorconfig-core-test as a submodule, but all other scripts IMO are integrated part of this subproject. Users in general get testing scripts, same for most other cases if they are supposed to get things from source. In addition, test scripts are unlikely to use a lot of spaces anyway.

cxw42 commented:
@xuhdev OK - I'll update this repo.  I may ask to revisit this point later if Bram decides he's willing to bundle this plugin with Vim.

cxw42 commented:
@xuhdev As of https://github.com/editorconfig/editorconfig-vim/pull/119/commits/e2bd6b86bacfefddbfb1f104f665824c9d4ce1ed , Travis tests both the core and the plugin.  I also have Appveyor testing the core so we have a Windows test case.  I cannot yet test the plugin on Windows since vimrunner doesn't support Windows (AndrewRadev/vimrunner#32).

xuhdev commented on README.md:
```diff
@@ -35,6 +28,10 @@ OR
 
   Then remember to call `:PluginInstall`.
 
+### No editorconfig core required!
```
Perhaps "No external editorconfig core library is required"?

xuhdev commented on autoload/editorconfig_core.vim:
```diff
@@ -0,0 +1,138 @@
+" autoload/editorconfig_core.vim: top-level functions for
+" editorconfig-core-vimscript.
+" Copyright (c) 2018 Chris White.  All rights reserved.
```
Can you add a license header block for every file? You can refer to other files in the plugin for how to do it. Also, should the year be 2019?

xuhdev commented on autoload/editorconfig_core.vim:
```diff
@@ -0,0 +1,138 @@
+endif
+" }}}1
+
+" The version we are, i.e., the latest version we support
```
This comment looks a bit weird. If it's about the latest version of the format we support, we should only have two numbers, because the third is for software update (bug fixes). I would just simply say "the version," because that's the least confusing and people should understand what it means.

cxw42 replied:
I am happy to change it.  Is there a wiki page or other document defining the project's versioning conventions?  I see that v0.12.2 is the PyPI version of editorconfig-core-py.  I also see that v0.12 is the latest version of editorconfig-core-test.  Does this mean that plugins should use the core-test version as their MAJOR.MINOR, but have their own PATCH?

The [core tests](https://github.com/editorconfig/editorconfig-core-test/blob/master/cli/CMakeLists.txt#L31) require the core to return a three-component version number.  Therefore, I have to use at least `0.12.0` to pass the tests.

What is the plan for the version numbering?  Thanks for helping me understand!

cxw42 replied:
@xuhdev Can we proceed with this while waiting for the general discussion in editorconfig issue 395?  vimcore currently passes all tests, including the version-related ones, and I believe this is the last open issue.  What do you think?

xuhdev replied:
Sorry, for some reason I didn't receive any notification about this. Perhaps it is the comment here that confused me ("the latest version we support" rather than "the version"). Which one do you intend it to be?

cxw42 replied:
@xuhdev I just updated my comment in editorconfig 395, and I updated the comment here in 53d117a7db6ba2c3da9efcf58b0734d011aadc37.  This is intended to be the specification version.  OK?

cxw42 replied:
Marked as resolved per your comment in https://github.com/editorconfig/editorconfig-vim/pull/119#pullrequestreview-238690349

xuhdev commented on autoload/editorconfig_core/fnmatch.vim:
```diff
@@ -0,0 +1,441 @@
+" encoding is utf8 - 中文
```
Is there a reason to have Chinese characters on this line?

cxw42 replied:
It is my personal style, but I can remove it if you wish.  I like to include some non-ASCII characters on the first line so that programs can more easily tell it's UTF-8 text.

xuhdev replied:
How about [IDEOGRAPHIC FULL STOP](http://www.fileformat.info/info/unicode/char/3002/index.htm)? It just looks semantically less confusing.

    encoding is utf8。

cxw42 replied:
Thanks!  I like that and have made the change.

xuhdev commented on autoload/editorconfig_core/ini.vim:
```diff
@@ -0,0 +1,252 @@
+
+"Based on code from ConfigParser.py file distributed with Python 2.6.
+
+"Licensed under PSF License (see LICENSE.PSF file).
```
We cannot license our files under PSF license, because PSF license is specifically made for Python Software Foundation. The first sentence of the PSF license reads

> This LICENSE AGREEMENT is between the Python Software Foundation ("PSF"), and the Individual or Organization ("Licensee") ...

cxw42 replied:
I updated the code to clarify.  We are not licensing under PSF.  However, we derived from PSF-licensed code, so have to include the PSF license text and copyright notice.

xuhdev commented on doc/editorconfig.txt:
```diff
@@ -8,16 +8,16 @@ Description: EditorConfig vim plugin
 License:
    Copyright (c) 2011-2015 EditorConfig Team
```
Can you also update the year?

cxw42 replied:
Sure!

cxw42 commented:
Question: what should the new version number be?  This plugin is currently v0.3.3.  Are we using semver? 
 Since we are removing functions, I think it should be v1.0.0 if we are using semver.  What do you think?  I will update the docs accordingly.

xuhdev commented:
Yes, I think v1.0.0 is the best new version number.

xuhdev commented:
Perhaps v0.99? Let's mark some instability for now.

cxw42 commented:
@xuhdev If we're actually semver, it could also be version `1.0.0-beta` to mark it as a prerelease.  OK?  If not, I'll use 0.99.

cxw42 commented:
Separate question - what is [`mkzip.sh`](https://github.com/editorconfig/editorconfig-vim/blob/master/mkzip.sh) used for?  I just noticed it for the first time - does it need to be updated?  It doesn't have any comments explaining its purpose.  However, it does mention Python, so I think it should be updated.

xuhdev commented:
Sounds good to me.

cxw42 commented:
@xuhdev I am logging off and will take a look at your next comments in the next few days.  Thanks for your detailed review!

xuhdev commented on autoload/editorconfig_core.vim:
```diff
@@ -0,0 +1,162 @@
+endfunction "editorconfig_core#currbuf_cli
+
+" }}}1
+" === Caching (not yet used) ============================================ {{{1
```
What do you mean by "not yet used"? Can we simply remove these code if they half-baken? We can make another PR to track the enabling of caching.

cxw42 replied:
Removed in https://github.com/editorconfig/editorconfig-vim/pull/119/commits/6a229d07c6e72019c80cd165c7511c7672bafb8e

cxw42 commented:
@xuhdev https://github.com/editorconfig/editorconfig-vim/pull/119#pullrequestreview-238690349

> Regarding to your new changes, I've made another comment. You also seem to have changed EOL of a few files. Just wanna make sure you intended it. 

Would you please let me know which files you are looking at?  I just looked through `git diff upstream/master` and didn't see any EOL changes.  All files currently have Unix (LF) line endings, which I think is reasonable.  (Except for the `crlf.in` file from editorconfig-core-test, which needs to have CRLF to function! :) )

There are a number of renames.  For example, I moved `tests/` to `tests/plugin` so that I could create `tests/core`.  Is that perhaps what you are looking at?

xuhdev commented:
Perhaps I'm getting confused. Could you tell me, for example, what has changed in `tests/core/ecvimlib.ps1` since then? I've seen a whole file change [here](https://github.com/editorconfig/editorconfig-vim/commit/53d117a7db6ba2c3da9efcf58b0734d011aadc37#diff-71263ae416d5a3f441fb25172968cb4e).

cxw42 commented:
Ah!  Sure.  Compared to `master`, that is a new file.  Everything in`tests/core` is copied in from `editorconfig-core-vimscript` per your earlier request to combine the repos.  But I see that GitHub, at least on mobile, does not make that clear at all!

cxw42 commented:
@xuhdev I cleaned up and squashed some.  The CRLF files are now consistent throughout the commits in which they occur, and the `.editorconfig` file specifies those line ends appropriately.  The previous commits are still available in [cxw42/vimcore-old](https://github.com/cxw42/editorconfig-vim/tree/vimcore-old) if you want to compare.

xuhdev commented:
Thank you for your hard work!

cxw42 commented:
@xuhdev Happy to help!  

When you have a chance, would you please set up Appveyor for this repo under editorconfig and update README.md?  It currently points to my Appveyor account.  No hurry, but I think it would be a good idea to continue running those tests.

xuhdev commented:
Done!

xuhdev commented:
Sorry for the late discussion on this issue, but I realized that we should have retained the ability to use an external core library as an option (the one that calls the external editorconfig executable, not the Python builtin). Could you restore the availability of that option, if this is not too difficult? We have done something similar to Emacs before: https://github.com/editorconfig/editorconfig-emacs/pull/46#issuecomment-160042488 I'm OK with defaulting to the internal one, but an option to use an external one might still be important.
